### PR TITLE
adds script to import cctbx data to rs

### PR DESCRIPTION
**Pull request by dermen**

### CODE REVIEW

**Comment by dermen** [comment]
After thinking about it, the cctbx dependencies could be ignored, I think the refl tables can completely be accessed using msgpack

```

In [92]: import msgpack

In [93]: d = msgpack.load(open("idx-0040_integrated.refl", "rb"), strict_map_key=False)

In [94]: vals = d[2]['data']['intensity.sum.value'][1]

In [95]: np.frombuffer(vals[1])
Out[95]: 
array([ 80.41394043, 648.45092773,  21.55432129, ...,  49.80627441,
       -22.07312012, 432.39892578])
```

So a lightweight version of this code could be written and not need any fancy dials conda env

**Comment by dermen** [comment]
Removed the CCTBX deps , slightly less functionality, but better performance

```bash
conda create -n cctbx2rs python=3.12
conda activate cctbx2rs
git clone https://github.com/dermen/reciprocalspaceship.git
cd reciprocalspaceship
pip install -e .

# new deps:
pip install ray msgpack

# test
wget https://smb.slac.stanford.edu/~dermen/rs_ints.tar
tar -xvf ts_ints.tar
rs.cctbx2rs  lys_si_8/ lys_si_8.mtz --ucell 78.3 78.3 36.9 90 90 90 --symbol P4 --nj 10
rs.mtzdump lys_si_8.mtz 

```

**Comment by kmdalton** [comment]
this is some inspiring stuff, @dermen. as we discussed on slack, i think rs-booster is the place for command line apps like this. however, i'd like to see this refactored into an `io` module so we can just run something like

```python
data_set = rs.read_cctbx( ... )
```

**Comment by JBGreisman** [comment]
Thanks for putting this together! I agree with Kevin here -- I think it would make sense to include an `rs.read_cctbx()` (or should it be called `rs.read_refl()`?). I think generic format conversion scripts are better suited for `rs-booster`, although such a commandline tool could then just call to the `rs.read....()` method.

**Comment by dermen** [comment]
Sounds good, ill rejig it. And yea, `read_refl` or even `read_dials` might be more suitable.. technically the refl files are produced by the program dials in most cases..

**Comment by kmdalton** [comment]
yes -- but we probably want different methods for stills and rotation method

**Comment by dermen** [comment]
https://github.com/rs-station/rs-booster/pull/50

**Comment by dermen** [comment]
@kmdalton , just saw your message.. I think I committed too quickly :)

**Comment by kmdalton** [comment]
no problem, we can always refactor the ray stuff after this pr. i think we still haven't converged on a general policy for methods that rely on / benefit from optional dependencies.

**Review by kmdalton — CHANGES_REQUESTED** [review]
@dermen , this is really nice, readable code. Thank you! I have a few requests before we merge it though. Most importantly,
1) the `rs` maintainers try to adhere to the classic unix philosophy of not printing extraneous info to console unless the user opts in. So, please put all your `print` statements behind some kind of `verbose=False` toggle. 
2) Your methods would be more flexible if they took filenames rather than directories. As written, it seems it would be impossible to only load a subset of the `.refl` files in a directory. This is an important feature to support. So, I'd request that you refactor all the "globbing" outside of the `read_...` methods. 
3) We have decorators (`from reciprocalspaceship.decorators import cellify, spacegroupify`) which take a range of possible input types and convert them to gemmi objects. Using these might simplify your code somewhat. 

I hope these comments amount to some light refactoring. I'm happy to help if needed!

After we merge this PR, I can refactor it to use contextmanagers as we discussed.

File: setup.py
Comment (dermen) [reply setup.py:91]: weird!

**Comment by kmdalton** [comment]
@dermen, I like the logging changes you made. I think the code doesn't suffer at all in readability, and it might be a good practice to deploy more widely in the `rs` codebase. 

there are a few more things i'd like to request with this pr. 
1) We typically make io functions available at the `rs.io.{function_name` level using the module's [\_\_init__.py](https://github.com/rs-station/reciprocalspaceship/blob/main/reciprocalspaceship/io/__init__.py). It'd be great if `rs.io.read_dials_stills` was populated in there. because you have optional dependencies, i think it'd be best if you 
2) implement a wrapper function with the signature `rs.io.read_dials_stills(fnames, unitcell, spacegroup, numjobs=10, backend="...")` where backend takes a string like `"mpi"` or`"ray"`. then make the default backend be `"numpy"` which doesn't require any additional libraries and runs in serial. 
3) Finally, it'd be good if `ray` and `mpi4py` were imported lazily and if the module could check whether they're available and fallback to `numpy` if not. i did something like this in the crystfel io code [here](https://github.com/rs-station/reciprocalspaceship/blob/1979862b14b0c41d389d8d769d1facf88af3b046/reciprocalspaceship/io/crystfel.py#L305). 

let me know if you need a hand with any of this. i can probably help out this week. thanks!

**Comment by dermen** [comment]
Cool - think its mostly there ! feel free to change things as you see fit! There is no real urgency for me with this.. I am actually anticipating we need to add more options for msgpack data columns, and try/excepts where a column name doesnt exist ...

File: setup.py
Comment (dermen) [reply setup.py:91]: Whats going on with this ? If you look in the dermen:main repo, you will see this line exists?

**Comment by dermen** [comment]
last commit was necessary to decorate `read_dials_stills`

**Comment by dermen** [comment]
Tests are in

**Comment by kmdalton** [comment]
i did some refactoring to de-emphasize the mpi idioms and make it possible to test the mpi code serially. the branch lives [here](https://github.com/dermen/reciprocalspaceship/compare/main...kmdalton:reciprocalspaceship:stills).

File: reciprocalspaceship/io/common.py
Comment (kmdalton) [inline reciprocalspaceship/io/common.py:21]: Oops this error message is now too specific. Change this to "Falling back to serial."

**Review by kmdalton — APPROVED** [review]
@dermen please review my changes and if you find them acceptable go ahead and merge this into main.

**Review by dermen — COMMENTED** [review]
I guess we still infer the stddev dtype ? Is that ok , and should we mention in docstring ?

File: reciprocalspaceship/io/dials.py
Comment (dermen) [inline reciprocalspaceship/io/dials.py:231]: did a word get left out here ?

**Comment by kmdalton** [comment]
> I guess we still infer the stddev dtype ? Is that ok , and should we mention in docstring ?

Ultimately that one is up to you and what you want the API to do. I would probably defer that, because it will downcast that column to float32. If we are going to infer the stddev dtype, we might as well infer things like HKL dtypes and BATCH dtype. Let me know what you think.

**Comment by kmdalton** [comment]
I added an option to specify whether to use mtz dtype inference.

**Comment by dermen** [comment]
Ok, fixed one syntax error and added a unit cell / space group estimator to  to rs-booster. And tested, seems to work ok:

```
$ rs.print_refl ml_proc.lysozyme_ML_mode_9/idx-0024_integrated.refl 

Found 8 experiment identifiers in ml_proc.lysozyme_ML_mode_9/idx-0024_integrated.refl:
	0: 17da648c-b938-5107-f582-7f7da8b803e8
	1: 1dccff2f-c147-29c6-55e3-163752dfb8fc
	2: 831a07f3-1af2-c739-ef9a-3c249f58e653
	3: 65853d5f-8458-d5e7-43a7-5101c5c3ee09
	4: 229d0132-1cb9-00ff-536a-6f872fede4dc
	5: f619a467-0abc-a18b-cbdf-75eef2df1f3e
	6: cde7c534-73ee-0d7b-5a39-77db7e62ee03
	7: 5b700c75-fa32-a892-0fdb-e53294402c3c

Reflection contents:
             names                           dtypes            
        background.dispersion                        double    
              background.mean                        double    
               background.mse                        double    
         background.sum.value                        double    
      background.sum.variance                        double    
                         bbox                          int6    
                            d                        double    
                delpsical.rad                        double    
                     entering                          bool    
                        flags                   std::size_t    
                           id                           int    
          intensity.sum.value                        double    
       intensity.sum.variance                        double    
                 miller_index        cctbx::miller::index<>    
        num_pixels.background                           int    
   num_pixels.background_used                           int    
        num_pixels.foreground                           int    
             num_pixels.valid                           int    
                        panel                   std::size_t    
                   partial_id                   std::size_t    
                   partiality                        double    
                           s1                  vec3<double>    
                      shoebox                     Shoebox<>    
                    xyzcal.mm                  vec3<double>    
                    xyzcal.px                  vec3<double>    
              xyzobs.mm.value                  vec3<double>    
           xyzobs.mm.variance                  vec3<double>    
              xyzobs.px.value                  vec3<double>    
           xyzobs.px.variance                  vec3<double>    

Number of reflections: 10555 


$ rs.from_dials ml_proc.lysozyme_ML_mode_9/ test.mtz --extra-cols num_pixels.foreground background.mean --verbose
Found 63 reflection files.
Found 63 corresponding exptlist files. Will use 5 of them to estimate ucell and space group.
Consensus unit cell from 35 expts: <gemmi.UnitCell(78.3001, 78.3001, 36.8998, 90, 90, 90)>
Consensus space group from 35 expts: <gemmi.SpaceGroup("P 4/m")>
2024-09-20 09:36:25,931	INFO worker.py:1781 -- Started a local Ray instance.
(_get_refl_data pid=2408434) Loading ml_proc.lysozyme_ML_mode_9/idx-0021_integrated.refl
(_get_refl_data pid=2408434) ... Read in data for miller_index
(_get_refl_data pid=2408434) ... Read in data for intensity.sum.value
(_get_refl_data pid=2408434) ... Read in data for intensity.sum.variance
(_get_refl_data pid=2408434) ... Read in data for xyzcal.px
(_get_refl_data pid=2408434) ... Read in data for s1
(_get_refl_data pid=2408434) ... Read in data for delpsical.rad
(_get_refl_data pid=2408434) ... Read in data for id
(_get_refl_data pid=2408434) ... Read in data for num_pixels.foreground
(_get_refl_data pid=2408434) ... Read in data for background.mean
(_get_refl_data pid=2408430) Loading ml_proc.lysozyme_ML_mode_9/idx-0051_integrated.refl [repeated 62x across cluster] (Ray deduplicates logs by default. Set RAY_DEDUP_LOGS=0 to disable log deduplication, or see https://docs.ray.io/en/master/ray-observability/user-guides/configure-logging.html#log-deduplication for more options.)
(_get_refl_data pid=2408444) ... Read in data for miller_index [repeated 62x across cluster]
(_get_refl_data pid=2408444) ... Read in data for intensity.sum.value [repeated 62x across cluster]
(_get_refl_data pid=2408444) ... Read in data for intensity.sum.variance [repeated 62x across cluster]
(_get_refl_data pid=2408444) ... Read in data for xyzcal.px [repeated 62x across cluster]
(_get_refl_data pid=2408444) ... Read in data for s1 [repeated 62x across cluster]
(_get_refl_data pid=2408444) ... Read in data for delpsical.rad [repeated 62x across cluster]
(_get_refl_data pid=2408444) ... Read in data for id [repeated 62x across cluster]
(_get_refl_data pid=2408444) ... Read in data for num_pixels.foreground [repeated 62x across cluster]
(_get_refl_data pid=2408444) ... Read in data for background.mean [repeated 62x across cluster]
Combining and formatting tables!
Found 592100 refls from 460 expts.
Mapping batch column.
Finished combining tables!
Converted column intensity.sum.variance to MTZ-Type Q, took sqrt of the values, and renamed to intensity.sum.sigma.
Writing MTZ test.mtz ...
Done writing MTZ.


$ rs.mtzdump  test.mtz 
Spacegroup: P4/m
Extended Hermann-Mauguin name: P 4/m
Unit cell dimensions: 78.300 78.300 36.900 90.000 90.000 90.000

mtz.head():

           background.mean  num_pixels.foreground  delpsical.rad  ...  intensity.sum.value  BATCH  PARTIAL
H   K  L                                                          ...                                     
-36 -6 21            2.146                    843         -0.002  ...               30.927     26     True
    -5 21            2.112                    842         -0.001  ...              -17.545     26     True
    -4 21            2.078                    837          0.001  ...               23.911     26     True
    -3 21            2.065                    838          0.002  ...                55.26     26     True
    -2 21            2.002                    827          0.002  ...              151.213     26     True

[5 rows x 13 columns]

mtz.describe():

       background.mean  num_pixels.foreground  delpsical.rad  ...  intensity.sum.sigma  intensity.sum.value     BATCH
count         592100.0               592100.0       592100.0  ...             592100.0             592100.0  592100.0
mean             10.05                404.062            0.0  ...               72.887             3568.129   230.097
std              7.908                143.466          0.003  ...               43.961            14724.827   133.145
min              0.017                  176.0         -0.046  ...                5.319            -7621.242       0.0
25%              3.714                  287.0         -0.002  ...               47.517               72.518     114.0
50%              6.843                  377.0            0.0  ...               59.611              293.854     231.0
75%             14.711                  494.0          0.002  ...               81.446             1530.024     345.0
max             50.248                 1106.0          0.043  ...              910.503           819365.438     459.0

[8 rows x 12 columns]

mtz.dtypes:

background.mean            MTZReal
num_pixels.foreground       MTZInt
delpsical.rad              MTZReal
s1.0                       MTZReal
s1.1                       MTZReal
s1.2                       MTZReal
xyzcal.px.0                MTZReal
xyzcal.px.1                MTZReal
xyzcal.px.2                MTZReal
intensity.sum.sigma         Stddev
intensity.sum.value      Intensity
BATCH                        Batch
PARTIAL                       bool
dtype: object



$ careless mono --kl-weight=0.001  --mc-samples=20  --mlp-layers=20  --image-layers=1 --dmin=2  --intensity-key=intensity.sum.value  --uncertainty-key=intensity.sum.sigma  dHKL,xyzcal.px.0,xyzcal.px.1,delpsical.rad,BATCH,background.mean,num_pixels.foreground   test.mtz  CL/test
Careless version 0.4.4
Training:   0%|                                                                                | 0/10000 [00:00<?, ?it/s]WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1726850289.048002  571008 service.cc:146] XLA service 0x7fd217b60350 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1726850289.048049  571008 service.cc:154]   StreamExecutor device (0): NVIDIA A100-SXM4-40GB, Compute Capability 8.0
I0000 00:00:1726850289.182732  571008 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.
Training:  13%|▎ | 1338/10000 [00:32<02:30, 57.64it/s, Grad Norm=1.50e+03, loss=1.27e+03, F KLDiv=1.13e+00, NLL=1.27e+03]

**Pull request merged**
